### DISCUSSION TITLE
Does Signaldb works on server side ?

### DISCUSSION THREAD

**Discussion by RSelwa** [maxnowack/signaldb#1060] [Q&A] [answered]
Hello, we're wondering if signal work on server side for mocking server response. Do you know if it's possible ? We didn't find the response in the docs. Thanks in advance

**Comment by maxnowack** [comment]
In theory all of the code also works on server side. Make sure that you use a persistence adapter that doesn't have browser dependencies (local storage for instance).
If it make sense to use signaldb on server, it depends on what to you want to achieve. I don't know what you exactly by mocking server response. Maybe you could explain it.

**Reply by RSelwa** [reply]
Understood.
Here is what we are trying to do: We have a Nextjs app, we want to use signalDb to mock our api response in e2e tests (Cypress here).
We've got two possible cases: client side and server side requests, so in client side, we can intercept the request and access to the data through Signaldb without having to make request to our firestore database. And we were wondering if we can have the same logic but server side but the persistence adapter we're using, for Firestore (syncManager), it seems that it use localstorage, so it doesn't work for server side.
Maybe we didn't find the correct way to use it.

**Reply by maxnowack** [reply, marked as answer]
You don't have to use the localStorage adapter with the `SyncManager`. You can pass a function to create your own persistence adapter instances to the constructor (see https://signaldb.js.org/sync/reference/#parameters).

On server side you could use the [file system](https://signaldb.js.org/data-persistence/file-system/) or also something like an in-memory adapter like we use in some unit tests:
https://github.com/maxnowack/signaldb/blob/main/packages/signaldb/__tests__/helpers/memoryPersistenceAdapter.ts

**Reply by RSelwa** [reply]
Thanks, it will be useful for us. last thing that isn't implemented yet is the asynchronous collections. With Cypress or another tool, we didn't find out how to wait for the collections to be created in signal and then replicated to the database. For now, we put explicit wait in the test code but it's not optimal you know 😅 . 
In one side, App -> Signal, there almost everytime a change in the interface so we can wait this change but from Signal to database, we don't have any solutions by now.
If you figure out how to make the collections asynchronous, Signal will be such an amazing tools 🤩

**Reply by maxnowack** [reply]
You could wait until `SyncManager.isReady()` to ensure that the SyncManager itself was loaded. On the collection you should wait until the event `persistence.init` was fired. After that you could execute `SyncManager.sync` to ensure the data is  up to date:

```js
await syncManager.isReady()
await new Promise(resolve => collection.once('persistence.init', resolve))
await syncManager.sync(collectionName)
```

**Comment by nickchomey** [comment]
@RSelwa did you ever get this working on the server?

**Reply by RSelwa** [reply]
Yes, we figured out to implement it server and client side 🎉

**Reply by nickchomey** [reply]
Great! Would you be able to share any details on how to do it?
